streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -0,0 +1,792 @@
+                            segment.put(key, segmentValue.serialize());
+                        }
+                    }
+                }
+
+                // update latest value store
+                if (value != null) {
+                    versionedStoreClient.putLatestValue(key, latestValueSchema.from(value, timestamp));
+                } else {
+                    versionedStoreClient.deleteLatestValue(key);
+                }
+                return new PutStatus(true, foundTs);
+            }
+        }
+        return new PutStatus(false, foundTs);
+    }
+
+    private static <T extends VersionedStoreSegment> PutStatus putSegments(
+        final SegmentValueSchema segmentValueSchema,
+        final VersionedStoreClient<T> versionedStoreClient,
+        final Optional<Sensor> expiredRecordSensor,
+        final ProcessorContext context,
+        final long observedStreamTime,
+        final long historyRetention,
+        final Bytes key,
+        final byte[] value,
+        final long timestamp,
+        final long prevFoundTs
+    ) {
+        // initialize with current foundTs value
+        long foundTs = prevFoundTs;
+
+        final List<T> segments = versionedStoreClient.getReverseSegments(timestamp);
+        for (final T segment : segments) {
+            final byte[] segmentValue = segment.get(key);
+            if (segmentValue != null) {
+                final long foundNextTs = segmentValueSchema.getNextTimestamp(segmentValue);
+                if (foundNextTs <= timestamp) {
+                    // this segment (and all earlier segments) does not contain records affected by
+                    // this put. insert into the segment specified by foundTs (i.e., the next
+                    // phase of the put() procedure) and conclude the procedure.
+                    return new PutStatus(false, foundTs);
+                }
+
+                final long foundMinTs = segmentValueSchema.getMinTimestamp(segmentValue);
+                if (foundMinTs <= timestamp) {
+                    // the record being inserted belongs in this segment.
+                    // insert and conclude the procedure.
+                    final long segmentIdForTimestamp = versionedStoreClient.segmentIdForTimestamp(timestamp);
+                    final boolean writeToOlderSegmentNeeded = segmentIdForTimestamp != segment.id();
+
+                    final SegmentValue sv = segmentValueSchema.deserialize(segmentValue);
+                    final SegmentSearchResult searchResult = sv.find(timestamp, writeToOlderSegmentNeeded);
+
+                    if (searchResult.validFrom() == timestamp) {
+                        // this put() replaces an existing entry, rather than adding a new one
+                        sv.updateRecord(timestamp, value, searchResult.index());
+                        segment.put(key, sv.serialize());
+                    } else {
+                        if (writeToOlderSegmentNeeded) {
```

Comment:
The control flow is hard to follow due to nesting... What about this:

(1) rename `writeToOlderSegmentNeeded` to `includeValue` (and explain in a comment why we need to include it.
(2) return directly inside if-blocks to avoid `else` and nesting:
```
if (searchResult.validFrom() == timestamp) {
  ...
  return PutStatus(true, foundTs);
}

if (includeValue) { // we included a value and thus need to move to older segment plus update found segment

  return PutStatus(true, foundTs);
}

// just plain insert into found segment
sv.insert(timestamp, value, searchResult.index());
segment.put(key, sv.serialize());
return PutStatus(true, foundTs);
```

Reply:
Sure, I'll make the changes.

My preference is to name the variable `writeToOlderSegmentMaybeNeeded` rather than `includeValue` since I think it's less natural (/ harder to explain) why we use a variable named `includeValue` to decide whether to write to an older segment or not, than it is to explain why we used a variable named `writeToOlderSegmentMaybeNeeded` to decide whether to include a value while searching. I'll go with that for now but can easily update it if you think `writeToOlderSegmentMaybeNeeded` with generous comments is still confusing.